# Term Project: End-to-End Cloud Data Pipeline
### Dataset: Student-chosen from Kaggle (with instructor approval).
Team project integrating Units 1–3 concepts.

## Project Requirements:
1. Define a clear business problem.
2. Batch Pipeline (Cloud Function + Scheduler) – automated data ingestion.
3. Streaming Pipeline (Pub/Sub + Dataflow) – real-time ingestion.
4. BigQuery ML – build and evaluate a predictive model.
5. Visualization – Looker Studio dashboard with KPIs.
6. Documentation – architecture diagram, decisions, prompts, lessons learned.

## Deliverables:
- Team notebook with all steps.
- Architecture blueprint (PDF or image).
- Screenshots of pipeline components running in GCP.
- Final Looker Studio dashboard link.

Rubric (Team 40%, Individual 60%):
- Data ingestion pipelines: 20%
- Streaming integration: 20%
- ML component: 20%
- Dashboard + KPIs: 20%
- Documentation & clarity: 20%
- Extra Credit (up to 10%): Innovation, additional datasets, advanced ML.


In [2]:
# Prompt output
from google.colab import auth
auth.authenticate_user() # Authenticates the Colab environment to Google Cloud

import os
PROJECT_ID = input("Enter your GCP Project ID: ").strip() # Prompt for and store the GCP Project ID
REGION = "us-central1"  # Set the default region; change if necessary

os.environ["GOOGLE_CLOUD_PROJECT"] = PROJECT_ID # Export the Project ID as an environment variable

# Set the active project for gcloud commands
!gcloud config set project $GOOGLE_CLOUD_PROJECT
print("Project:", PROJECT_ID, "| Region:", REGION)


Enter your GCP Project ID: sunlit-plasma-471119-s7
Updated property [core/project].
Project: sunlit-plasma-471119-s7 | Region: us-central1


In [3]:
from google.colab import files
print("Upload your kaggle.json (Kaggle > Account > Create New API Token)")
uploaded = files.upload()

import os
os.makedirs('/root/.kaggle', exist_ok=True)
with open('/root/.kaggle/kaggle.json', 'wb') as f:
    f.write(uploaded[list(uploaded.keys())[0]])
os.chmod('/root/.kaggle/kaggle.json', 0o600)  # owner-only permissions for security

!kaggle --version

Upload your kaggle.json (Kaggle > Account > Create New API Token)


Saving kaggle.json to kaggle.json
Kaggle API 1.7.4.5


In [4]:
# Create directory for raw data
!mkdir -p /content/data/raw

# Download the dataset using Kaggle CLI
!kaggle datasets download -d usgs/earthquake-database -p /content/data

# Unzip the downloaded file into the raw data directory (overwrite if necessary)
!unzip -o /content/data/*.zip -d /content/data/raw

# List all CSV files in the raw data directory with their sizes
!ls -lh /content/data/raw/*.csv

Dataset URL: https://www.kaggle.com/datasets/usgs/earthquake-database
License(s): CC0-1.0
  0% 0.00/590k [00:00<?, ?B/s]
100% 590k/590k [00:00<00:00, 908MB/s]
Archive:  /content/data/earthquake-database.zip
  inflating: /content/data/raw/database.csv  
-rw-r--r-- 1 root root 2.3M Sep 20  2019 /content/data/raw/database.csv


In [5]:
import uuid
import os

# Create a unique bucket name with a random suffix
bucket_name = f"mgmt467-earthquake-{uuid.uuid4().hex[:8]}"
os.environ["BUCKET_NAME"] = bucket_name

# Create the GCS bucket
# Note: If you encounter an error related to the region,
# you might need to check available locations with '!gcloud storage locations list'
# and potentially update the REGION variable at the top of the notebook.
!gcloud storage buckets create gs://$BUCKET_NAME --location=US

# Upload all CSV files to the bucket under the 'netflix/' prefix
!gcloud storage cp /content/data/raw/* gs://$BUCKET_NAME/earthquake/

# Print the bucket name and explain the benefits of staging
print(f"Created bucket: {bucket_name}")
print("\nBenefits of staging data in GCS:")
print("- Provides a consistent and versionable source for data.")
print("- Decouples data storage from processing, allowing for flexible use with various GCP services like BigQuery.")
print("- Enables central management and access control of data.")

Creating gs://mgmt467-earthquake-11d76c04/...
Copying file:///content/data/raw/database.csv to gs://mgmt467-earthquake-11d76c04/earthquake/database.csv
Created bucket: mgmt467-earthquake-11d76c04

Benefits of staging data in GCS:
- Provides a consistent and versionable source for data.
- Decouples data storage from processing, allowing for flexible use with various GCP services like BigQuery.
- Enables central management and access control of data.


In [6]:
# Verification: List contents of the earthquake/ prefix in the bucket
import os
bucket_name = os.environ["BUCKET_NAME"]
!gcloud storage ls gs://$BUCKET_NAME/earthquake/ --recursive --readable-sizes

gs://mgmt467-earthquake-11d76c04/earthquake/:
gs://mgmt467-earthquake-11d76c04/earthquake/database.csv


In [7]:
# Create BigQuery dataset
DATASET = "earthquake"
# Attempt to create; ignore if exists
!bq --location=US mk -d --description "MGMT467 Earthquake dataset" $DATASET || echo "Dataset may already exist."

BigQuery error in mk operation: Dataset 'sunlit-plasma-471119-s7:earthquake'
already exists.
Dataset may already exist.


In [8]:
# Load table from GCS
tables = {
  "database": "database.csv",
}
import os
for tbl, fname in tables.items():
  src = f"gs://{os.environ['BUCKET_NAME']}/earthquake/{fname}"
  print("Loading", tbl, "from", src)
  !bq load --skip_leading_rows=1 --autodetect --source_format=CSV {DATASET}.{tbl} {src}

Loading database from gs://mgmt467-earthquake-11d76c04/earthquake/database.csv
Waiting on bqjob_r4a075ed5e135136c_0000019b1aae2347_1 ... (1s) Current status: DONE   


In [9]:
%%bigquery
SELECT
  *
FROM
  `sunlit-plasma-471119-s7.earthquake.database`
LIMIT 10;

Query is running:   0%|          |

Downloading:   0%|          |

Date      Time  Latitude  Longitude        Type  Depth  Depth Error  \
0  01/02/1965  13:44:18    19.246    145.616  Earthquake  131.6          NaN   
1  01/04/1965  11:29:49     1.863    127.352  Earthquake   80.0          NaN   
2  01/05/1965  18:05:58   -20.579   -173.972  Earthquake   20.0          NaN   
3  01/08/1965  18:49:43   -59.076    -23.557  Earthquake   15.0          NaN   
4  01/09/1965  13:32:50    11.938    126.427  Earthquake   15.0          NaN   
5  01/10/1965  13:36:32   -13.405    166.629  Earthquake   35.0          NaN   
6  01/12/1965  13:32:25    27.357     87.867  Earthquake   20.0          NaN   
7  01/15/1965  23:17:42   -13.309    166.212  Earthquake   35.0          NaN   
8  01/16/1965  11:32:37   -56.452    -27.043  Earthquake   95.0          NaN   
9  01/17/1965  10:43:17   -24.563    178.487  Earthquake  565.0          NaN   

   Depth Seismic Stations  Magnitude Magnitude Type  ...  \
0                    <NA>        6.0             MW  ...   
1                    <NA>        5.8             MW  ...   
2                    <NA>        6.2             MW  ...   
3                    <NA>        5.8             MW  ...   
4                    <NA>        5.8             MW  ...   
5                    <NA>        6.7             MW  ...   
6                    <NA>        5.9             MW  ...   
7                    <NA>        6.0             MW  ...   
8                    <NA>        6.0             MW  ...   
9                    <NA>        5.8             MW  ...   

   Magnitude Seismic Stations  Azimuthal Gap  Horizontal Distance  \
0                        <NA>            NaN                  NaN   
1                        <NA>            NaN                  NaN   
2                        <NA>            NaN                  NaN   
3                        <NA>            NaN                  NaN   
4                        <NA>            NaN                  NaN   
5                        <NA>            NaN                  NaN   
6                        <NA>            NaN                  NaN   
7                        <NA>            NaN                  NaN   
8                        <NA>            NaN                  NaN   
9                        <NA>            NaN                  NaN   

   Horizontal Error  Root Mean Square               ID     Source  \
0               NaN               NaN     ISCGEM860706     ISCGEM   
1               NaN               NaN     ISCGEM860737     ISCGEM   
2               NaN               NaN     ISCGEM860762     ISCGEM   
3               NaN               NaN     ISCGEM860856     ISCGEM   
4               NaN               NaN     ISCGEM860890     ISCGEM   
5               NaN               NaN     ISCGEM860922     ISCGEM   
6               NaN               NaN     ISCGEM861007     ISCGEM   
7               NaN               NaN     ISCGEM861111     ISCGEM   
8               NaN               NaN  ISCGEMSUP861125  ISCGEMSUP   
9               NaN               NaN     ISCGEM861148     ISCGEM   

  Location Source Magnitude Source     Status  
0          ISCGEM           ISCGEM  Automatic  
1          ISCGEM           ISCGEM  Automatic  
2          ISCGEM           ISCGEM  Automatic  
3          ISCGEM           ISCGEM  Automatic  
4          ISCGEM           ISCGEM  Automatic  
5          ISCGEM           ISCGEM  Automatic  
6          ISCGEM           ISCGEM  Automatic  
7          ISCGEM           ISCGEM  Automatic  
8          ISCGEM           ISCGEM  Automatic  
9          ISCGEM           ISCGEM  Automatic  

[10 rows x 21 columns]

In [10]:
import pandas as pd
import pandas_gbq

project_id = os.environ["GOOGLE_CLOUD_PROJECT"]
table_id = f"{project_id}.{DATASET}.database"

# Load data from BigQuery into a pandas DataFrame
df = pandas_gbq.read_gbq(f"SELECT * FROM `{table_id}`", project_id=project_id)

print("Shape of DataFrame before dropping NA values:", df.shape)

Downloading: 100%|██████████|
Shape of DataFrame before dropping NA values: (46824, 21)


In [11]:
df.isna().sum()

Date                              0
Time                              0
Latitude                          0
Longitude                         0
Type                              0
Depth                             0
Depth Error                   37902
Depth Seismic Stations        32630
Magnitude                         0
Magnitude Type                    6
Magnitude Error               46170
Magnitude Seismic Stations    41696
Azimuthal Gap                 32226
Horizontal Distance           43616
Horizontal Error              44512
Root Mean Square              12120
ID                                0
Source                            0
Location Source                   0
Magnitude Source                  0
Status                            0
dtype: int64

In [12]:
df["Depth Seismic Stations"].value_counts()
df = df.drop(columns=["Magnitude Seismic Stations", "Magnitude Error", "Horizontal Error", "Root Mean Square", "Horizontal Distance", "Azimuthal Gap", "Depth Error", "Depth Seismic Stations"])

In [15]:
# Drop rows with any NA values
df_cleaned = df.dropna()

print("Shape of DataFrame after dropping NA values:", df_cleaned.shape)
display(df_cleaned.head())
print(df_cleaned.shape)

Shape of DataFrame after dropping NA values: (46818, 13)


Date      Time  Latitude  Longitude        Type  Depth  Magnitude  \
0  01/02/1965  13:44:18    19.246    145.616  Earthquake  131.6        6.0   
1  01/04/1965  11:29:49     1.863    127.352  Earthquake   80.0        5.8   
2  01/05/1965  18:05:58   -20.579   -173.972  Earthquake   20.0        6.2   
3  01/08/1965  18:49:43   -59.076    -23.557  Earthquake   15.0        5.8   
4  01/09/1965  13:32:50    11.938    126.427  Earthquake   15.0        5.8   

  Magnitude Type            ID  Source Location Source Magnitude Source  \
0             MW  ISCGEM860706  ISCGEM          ISCGEM           ISCGEM   
1             MW  ISCGEM860737  ISCGEM          ISCGEM           ISCGEM   
2             MW  ISCGEM860762  ISCGEM          ISCGEM           ISCGEM   
3             MW  ISCGEM860856  ISCGEM          ISCGEM           ISCGEM   
4             MW  ISCGEM860890  ISCGEM          ISCGEM           ISCGEM   

      Status  
0  Automatic  
1  Automatic  
2  Automatic  
3  Automatic  
4  Automatic

(46818, 13)


### Preprocessing Data for BigQuery ML

We will prepare the `df` DataFrame for BigQuery ML. This involves:
1.  Combining 'Date' and 'Time' columns into a single datetime object.
2.  Extracting new features like 'Year', 'Month', 'Day', and 'Hour' from the datetime.
3.  Dropping the original 'Date' and 'Time' columns.
4.  Handling the remaining 3 missing values in 'Magnitude Type' by dropping those rows.
5.  Selecting relevant features for our prediction model.

In [18]:
import pandas as pd
import pandas_gbq

# Make a copy to avoid modifying the original df directly in case it's needed later
df_bqml = df_cleaned.copy()

# Combine 'Date' and 'Time' into a single datetime column
df_bqml['datetime'] = pd.to_datetime(df_bqml['Date'] + ' ' + df_bqml['Time'], errors='coerce')

# Drop rows where datetime conversion failed (if any)
df_bqml.dropna(subset=['datetime'], inplace=True)

# Extract temporal features
df_bqml['Year'] = df_bqml['datetime'].dt.year
df_bqml['Month'] = df_bqml['datetime'].dt.month
df_bqml['Day'] = df_bqml['datetime'].dt.day
df_bqml['Hour'] = df_bqml['datetime'].dt.hour

# Drop original 'Date', 'Time', and the temporary 'datetime' column
df_bqml = df_bqml.drop(columns=['Date', 'Time', 'datetime'])

# Handle remaining missing values in 'Magnitude Type' (only 3 NAs)
#df_bqml.dropna(subset=['Magnitude Type'], inplace=True)

# Select features for the BQML model (target is 'Magnitude')
# 'ID', 'Source', 'Location Source', 'Magnitude Source', 'Status' are metadata or source-related, not typically used as direct features for prediction in a barebones model.
features = [
    'Latitude',
    'Longitude',
    'Depth',
    'Type',
    'Magnitude Type',
    'Year',
    'Month',
    'Day',
    'Hour',
    'ID',
    'Source',
    'Location Source',
    'Magnitude Source',
    'Status',
    'Magnitude' # This is our target variable
]

df_bqml_processed = df_bqml[features]

print("Shape of DataFrame after preprocessing for BQML:", df_bqml_processed.shape)
display(df_bqml_processed.head())
df_bqml_processed.isna().sum()

Shape of DataFrame after preprocessing for BQML: (46812, 15)


Latitude  Longitude  Depth        Type Magnitude Type  Year  Month  Day  \
0    19.246    145.616  131.6  Earthquake             MW  1965      1    2   
1     1.863    127.352   80.0  Earthquake             MW  1965      1    4   
2   -20.579   -173.972   20.0  Earthquake             MW  1965      1    5   
3   -59.076    -23.557   15.0  Earthquake             MW  1965      1    8   
4    11.938    126.427   15.0  Earthquake             MW  1965      1    9   

   Hour            ID  Source Location Source Magnitude Source     Status  \
0    13  ISCGEM860706  ISCGEM          ISCGEM           ISCGEM  Automatic   
1    11  ISCGEM860737  ISCGEM          ISCGEM           ISCGEM  Automatic   
2    18  ISCGEM860762  ISCGEM          ISCGEM           ISCGEM  Automatic   
3    18  ISCGEM860856  ISCGEM          ISCGEM           ISCGEM  Automatic   
4    13  ISCGEM860890  ISCGEM          ISCGEM           ISCGEM  Automatic   

   Magnitude  
0        6.0  
1        5.8  
2        6.2  
3        5.8  
4        5.8

Latitude            0
Longitude           0
Depth               0
Type                0
Magnitude Type      0
Year                0
Month               0
Day                 0
Hour                0
ID                  0
Source              0
Location Source     0
Magnitude Source    0
Status              0
Magnitude           0
dtype: int64

### Upload Processed Data to BigQuery

Now we will upload the preprocessed DataFrame to a new BigQuery table, which will serve as the input for our BigQuery ML model.

In [32]:
# Define the new BigQuery table ID for the processed data
BQML_TABLE_NAME = "earthquake_bqml_data"
bqml_table_id = f"{PROJECT_ID}.{DATASET}.{BQML_TABLE_NAME}"

# Upload the processed DataFrame to BigQuery
pandas_gbq.to_gbq(
    df_bqml_processed,
    bqml_table_id,
    project_id=PROJECT_ID,
    if_exists='replace' # Replace table if it already exists
)

print(f"Processed data uploaded to BigQuery table: {bqml_table_id}")

100%|██████████| 1/1 [00:00<00:00, 8594.89it/s]

Processed data uploaded to BigQuery table: sunlit-plasma-471119-s7.earthquake.earthquake_bqml_data


In [33]:
#Check bigquery processed table
%%bigquery
SELECT
  *
FROM
  `sunlit-plasma-471119-s7.earthquake.earthquake_bqml_data`
LIMIT 10;

Query is running:   0%|          |

Downloading:   0%|          |

Latitude  Longitude  Depth        Type Magnitude Type  Year  Month  Day  \
0   67.7185  -162.6679   23.4  Earthquake             ML  2014      4   18   
1   67.7185  -162.6679   23.4  Earthquake             ML  2014      4   18   
2   60.0035  -152.1323   91.1  Earthquake             ML  2014      5   10   
3   60.0035  -152.1323   91.1  Earthquake             ML  2014      5   10   
4   67.7245  -162.3749   18.6  Earthquake             ML  2014      6    7   
5   67.7245  -162.3749   18.6  Earthquake             ML  2014      6    7   
6   67.6973  -162.6119   24.2  Earthquake             ML  2014      6   16   
7   67.6973  -162.6119   24.2  Earthquake             ML  2014      6   16   
8   61.9449  -151.8159  108.9  Earthquake             ML  2014      9   25   
9   61.9449  -151.8159  108.9  Earthquake             ML  2014      9   25   

   Hour          ID Source Location Source Magnitude Source    Status  \
0    18  AK11232962     AK              AK               AK  Reviewed   
1    18  AK11232962     AK              AK               AK  Reviewed   
2    14  AK11257368     AK              AK               AK  Reviewed   
3    14  AK11257368     AK              AK               AK  Reviewed   
4     4  AK11285090     AK              AK               AK  Reviewed   
5     4  AK11285090     AK              AK               AK  Reviewed   
6    12  AK11292524     AK              AK               AK  Reviewed   
7    12  AK11292524     AK              AK               AK  Reviewed   
8    17  AK11401855     AK              AK               AK  Reviewed   
9    17  AK11401855     AK              AK               AK  Reviewed   

   Magnitude  
0        5.6  
1        5.6  
2        5.6  
3        5.6  
4        5.5  
5        5.5  
6        5.7  
7        5.7  
8        6.2  
9        6.2

# Task
Create a new BigQuery table `earthquake.earthquake_bqml_enhanced_data` by selecting all columns from `sunlit-plasma-471119-s7.earthquake.earthquake_bqml_data` and adding new features: `sin_month`, `cos_month`, `sin_hour`, `cos_hour`, `latitude_x_longitude`, `depth_x_magnitude`, `depth_squared`, and `magnitude_squared`.

## Engineer Cyclical Time Features

### Subtask:
Create new features like `sin_month`, `cos_month`, `sin_hour`, and `cos_hour` from the existing `Month` and `Hour` columns using SQL in BigQuery. This helps the model understand the cyclical nature of time.


**Reasoning**:
The subtask requires creating cyclical features in BigQuery. I will use a `%%bigquery` magic command to execute a SQL query that creates a new table with the calculated `sin_month`, `cos_month`, `sin_hour`, and `cos_hour` columns, along with all existing columns from the processed data table.



In [34]:
%%bigquery
CREATE OR REPLACE TABLE `sunlit-plasma-471119-s7.earthquake.earthquake_bqml_enhanced_data` AS
SELECT
  *,
  SIN(2 * ACOS(-1) * Month / 12) AS sin_month,
  COS(2 * ACOS(-1) * Month / 12) AS cos_month,
  SIN(2 * ACOS(-1) * Hour / 24) AS sin_hour,
  COS(2 * ACOS(-1) * Hour / 24) AS cos_hour
FROM
  `sunlit-plasma-471119-s7.earthquake.earthquake_bqml_data`

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

**Reasoning**:
The previous step successfully created the BigQuery table with cyclical time features. Now, I need to add the remaining specified engineered features (`latitude_x_longitude`, `depth_squared`) to this table. I will use a `CREATE OR REPLACE TABLE` statement on the `earthquake_bqml_enhanced_data` table to achieve this, calculating the new features based on existing columns.



In [35]:
%%bigquery
CREATE OR REPLACE TABLE `sunlit-plasma-471119-s7.earthquake.earthquake_bqml_enhanced_data` AS
SELECT
  *,
  Latitude * Longitude AS latitude_x_longitude,
  POW(Depth, 2) AS depth_squared,
FROM
  `sunlit-plasma-471119-s7.earthquake.earthquake_bqml_enhanced_data`

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [36]:
%%bigquery
SELECT
  *
FROM
  `sunlit-plasma-471119-s7.earthquake.earthquake_bqml_enhanced_data`
LIMIT 10;

Query is running:   0%|          |

Downloading:   0%|          |

Latitude  Longitude  Depth        Type Magnitude Type  Year  Month  Day  \
0    19.250   -155.490   10.9  Earthquake            MWC  1999      4   17   
1    19.250   -155.490   10.9  Earthquake            MWC  1999      4   17   
2    42.126    135.008  395.0  Earthquake             MW  2009     12   24   
3    42.126    135.008  395.0  Earthquake             MW  2009     12   24   
4     6.405    126.640   60.0  Earthquake             MW  1972     12    2   
5     6.405    126.640   60.0  Earthquake             MW  1972     12    2   
6    44.639    149.276   39.0  Earthquake             MW  1972     12   17   
7    44.639    149.276   39.0  Earthquake             MW  1972     12   17   
8    -6.134    148.846   35.0  Earthquake             MW  1972     12   19   
9    -6.134    148.846   35.0  Earthquake             MW  1972     12   19   

   Hour              ID  ... Location Source Magnitude Source     Status  \
0     0    GCMTB041799A  ...            GCMT             GCMT  Automatic   
1     0    GCMTB041799A  ...            GCMT             GCMT  Automatic   
2     0  ISCGEM17576175  ...          ISCGEM           ISCGEM  Automatic   
3     0  ISCGEM17576175  ...          ISCGEM           ISCGEM  Automatic   
4     0    ISCGEM765501  ...          ISCGEM           ISCGEM  Automatic   
5     0    ISCGEM765501  ...          ISCGEM           ISCGEM  Automatic   
6     0    ISCGEM766233  ...          ISCGEM           ISCGEM  Automatic   
7     0    ISCGEM766233  ...          ISCGEM           ISCGEM  Automatic   
8     0    ISCGEM766360  ...          ISCGEM           ISCGEM  Automatic   
9     0    ISCGEM766360  ...          ISCGEM           ISCGEM  Automatic   

  Magnitude     sin_month  cos_month  sin_hour  cos_hour  \
0       5.7  8.660254e-01       -0.5       0.0       1.0   
1       5.7  8.660254e-01       -0.5       0.0       1.0   
2       5.7 -2.449294e-16        1.0       0.0       1.0   
3       5.7 -2.449294e-16        1.0       0.0       1.0   
4       8.0 -2.449294e-16        1.0       0.0       1.0   
5       8.0 -2.449294e-16        1.0       0.0       1.0   
6       6.4 -2.449294e-16        1.0       0.0       1.0   
7       6.4 -2.449294e-16        1.0       0.0       1.0   
8       5.8 -2.449294e-16        1.0       0.0       1.0   
9       5.8 -2.449294e-16        1.0       0.0       1.0   

   latitude_x_longitude  depth_squared  
0          -2993.182500         118.81  
1          -2993.182500         118.81  
2           5687.347008      156025.00  
3           5687.347008      156025.00  
4            811.129200        3600.00  
5            811.129200        3600.00  
6           6663.531364        1521.00  
7           6663.531364        1521.00  
8           -913.021364        1225.00  
9           -913.021364        1225.00  

[10 rows x 21 columns]

## Retrain BigQuery ML Model

### Subtask:
Train a new BigQuery ML linear regression model using the enhanced dataset, ensuring the model definition correctly references the new features and the updated table.


**Reasoning**:
I need to train a new BigQuery ML linear regression model using the enhanced dataset and the specified features.



# Task
Retrain the `earthquake_magnitude_enhanced_model` using a random 80/20 train-test split and then evaluate its performance on the held-out test set using `ML.EVALUATE`, finally reviewing the evaluation results to discuss its performance on unseen data.

## Retrain BigQuery ML Model with Train-Test Split

### Subtask:
Retrain the `earthquake_magnitude_enhanced_model` using an 80/20 train-test split, ensuring the dataset includes all specified enhanced features.


**Reasoning**:
Now that the `earthquake_bqml_enhanced_data` table has been updated with all the required engineered features, I will train a new BigQuery ML linear regression model using a random 80/20 train-test split as specified.



In [53]:
%%bigquery
CREATE OR REPLACE MODEL `sunlit-plasma-471119-s7.earthquake.earthquake_magnitude_model_final`
OPTIONS(
  model_type = 'BOOSTED_TREE_REGRESSOR',
  input_label_cols = ['Magnitude'],
  DATA_SPLIT_METHOD = 'RANDOM',
  data_split_eval_fraction = 0.2
)
AS
SELECT
  Latitude,
  Longitude,
  Depth,
  Type,
  `Magnitude Type` AS Magnitude_Type,
  Year,
  Month,
  Day,
  Hour,
  sin_month,
  cos_month,
  sin_hour,
  cos_hour,
  latitude_x_longitude,
  depth_squared,
  Magnitude
FROM
  `sunlit-plasma-471119-s7.earthquake.earthquake_bqml_enhanced_data`;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

**Reasoning**:
The model has been retrained with a train-test split. Now, I need to evaluate its performance on the held-out test set as per the instructions.



In [54]:
%%bigquery
SELECT
  *
FROM
  ML.EVALUATE(MODEL `sunlit-plasma-471119-s7.earthquake.earthquake_magnitude_model_final`)

Query is running:   0%|          |

Downloading:   0%|          |

mean_absolute_error  mean_squared_error  mean_squared_log_error  \
0             0.274443            0.152367                0.002884   

   median_absolute_error  r2_score  explained_variance  
0               0.206433  0.148614            0.156412

In [55]:
%%bigquery
CREATE OR REPLACE MODEL `sunlit-plasma-471119-s7.earthquake.earthquake_magnitude_model_linreg`
OPTIONS(
  model_type = 'LINEAR_REG',
  input_label_cols = ['Magnitude'],
  DATA_SPLIT_METHOD = 'RANDOM',
  data_split_eval_fraction = 0.2
)
AS
SELECT
  Latitude,
  Longitude,
  Depth,
  Type,
  `Magnitude Type` AS Magnitude_Type,
  Year,
  Month,
  Day,
  Hour,
  sin_month,
  cos_month,
  sin_hour,
  cos_hour,
  latitude_x_longitude,
  depth_squared,
  Magnitude
FROM
  `sunlit-plasma-471119-s7.earthquake.earthquake_bqml_enhanced_data`;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [56]:
%%bigquery
SELECT
  *
FROM
  ML.EVALUATE(MODEL `sunlit-plasma-471119-s7.earthquake.earthquake_magnitude_model_linreg`)

Query is running:   0%|          |

Downloading:   0%|          |

mean_absolute_error  mean_squared_error  mean_squared_log_error  \
0              0.29982            0.166587                0.003184   

   median_absolute_error  r2_score  explained_variance  
0               0.242225  0.069152            0.069153

## Model Comparison Summary

We have trained and evaluated two different BigQuery ML models on the `earthquake_bqml_enhanced_data` dataset, both using a random 80/20 train-test split. The models are:

1.  `earthquake_magnitude_model_final` (Boosted Tree Regressor)
2.  `earthquake_magnitude_model_linreg` (Linear Regression)

Let's compare their evaluation metrics on the held-out test set:

### Evaluation Results

| Metric                   | Boosted Tree Regressor (`earthquake_magnitude_model_final`) | Linear Regression (`earthquake_magnitude_model_linreg`) |
| :----------------------- | :--------------------------------------------------------- | :------------------------------------------------------ |
| **Mean Absolute Error**  | 0.274443                                                   | 0.29982                                                 |
| **Mean Squared Error**   | 0.152367                                                   | 0.166587                                                |
| **R2 Score**             | 0.148614                                                   | 0.069152                                                |
| **Explained Variance**   | 0.156412                                                   | 0.069153                                                |

### Discussion

From the comparison, the **Boosted Tree Regressor (`earthquake_magnitude_model_final`) clearly outperforms the Linear Regression model (`earthquake_magnitude_model_linreg`)** on this dataset. Here's why:

*   **Lower Errors**: The Boosted Tree Regressor shows lower values for both Mean Absolute Error (MAE) and Mean Squared Error (MSE), indicating that its predictions are, on average, closer to the actual earthquake magnitudes.
*   **Higher R2 Score**: The R2 Score for the Boosted Tree Regressor is significantly higher (0.1486) compared to the Linear Regression model (0.0691). The R2 score represents the proportion of variance in the dependent variable that can be predicted from the independent variables. A higher R2 score indicates a better fit of the model to the data. In this case, the Boosted Tree model explains almost double the variance in earthquake magnitudes compared to the Linear Regression model.
*   **Better Explained Variance**: Similarly, the Explained Variance score is higher for the Boosted Tree Regressor, reinforcing its superior ability to capture the underlying patterns in the data.

### Conclusion

The Boosted Tree Regressor is the better-performing model for predicting earthquake magnitudes with the current set of engineered features and data split. Its non-linear nature likely allows it to capture more complex relationships within the data that a simple linear model cannot. Further optimization could involve hyperparameter tuning for the Boosted Tree Regressor or exploring other advanced ensemble methods.

## DIVE (Discover, Evaluate, Validate, Extend) for Earthquake Magnitude Prediction

This DIVE framework outlines the approach taken and future considerations for the earthquake magnitude prediction task using BigQuery ML.

### 1. Discover

*   **Problem Statement**: Predict the `Magnitude` of an earthquake based on various seismic and temporal features.
*   **Data Source**: USGS Earthquake Database from Kaggle, loaded into BigQuery (`earthquake.database`, then preprocessed into `earthquake.earthquake_bqml_data`).
*   **Initial Features**: Latitude, Longitude, Depth, Type, Magnitude Type, Year, Month, Day, Hour, ID, Source, Location Source, Status.
*   **Feature Engineering Ideas**: Cyclical features for time (Month, Hour), interaction terms (Latitude x Longitude, Depth x Magnitude), and polynomial features (Depth^2, Magnitude^2).

### 2. Evaluate

*   **Target Metric**: Regression metrics such as Mean Absolute Error (MAE), Mean Squared Error (MSE), R-squared (R2 Score), and Explained Variance are used to evaluate model performance.
*   **Models Explored**:
    *   **Boosted Tree Regressor**: Chosen for its ability to handle non-linear relationships and potentially complex interactions.
    *   **Linear Regression**: Used as a baseline for comparison, representing a simpler, linear relationship.
*   **Current Performance (Boosted Tree Regressor)**: MAE = 0.274443, MSE = 0.152367, R2 Score = 0.148614.
*   **Current Performance (Linear Regression)**: MAE = 0.29982, MSE = 0.166587, R2 Score = 0.069152.

### 3. Validate

*   **Train-Test Split**: An 80/20 random train-test split was applied to both models to assess generalization performance on unseen data. This helps identify overfitting.
*   **Model Comparison**: Performance metrics were explicitly compared between the Boosted Tree Regressor and Linear Regression models to determine the more suitable algorithm for this problem.
*   **Feature Inclusion**: Ensured that all engineered features (cyclical, interaction, polynomial) were correctly included in the model training to assess their collective impact.

### 4. Extend

*   **Deployment**: Once a satisfactory model is achieved, consider deploying it for real-time predictions using BigQuery ML's `ML.PREDICT` function or integrating it into a larger data pipeline.
*   **Monitoring**: Implement continuous monitoring of model performance in a production environment to detect degradation over time and trigger retraining.

## Explain Predictions for the Boosted Tree Model

Using `ML.EXPLAIN_PREDICT` allows us to understand the contribution of each feature to a model's individual predictions. For tree-based models like Boosted Trees, this often uses SHAP (SHapley Additive exPlanations) values to attribute the prediction to each input feature.

In [59]:
%%bigquery
SELECT
  *
FROM
  ML.EXPLAIN_PREDICT(MODEL `sunlit-plasma-471119-s7.earthquake.earthquake_magnitude_model_final`,
    (SELECT
      Latitude,
      Longitude,
      Depth,
      Type,
      `Magnitude Type` AS Magnitude_Type,
      Year,
      Month,
      Day,
      Hour,
      sin_month,
      cos_month,
      sin_hour,
      cos_hour,
      latitude_x_longitude,
      depth_squared,
      Magnitude
    FROM
      `sunlit-plasma-471119-s7.earthquake.earthquake_bqml_enhanced_data`)
  )
LIMIT 10;

Query is running:   0%|          |

Downloading:   0%|          |

predicted_Magnitude                           top_feature_attributions  \
0             6.068205  [{'feature': 'Depth', 'attribution': 0.1112136...   
1             6.068205  [{'feature': 'Depth', 'attribution': 0.1112136...   
2             5.822044  [{'feature': 'Hour', 'attribution': -0.0287152...   
3             5.822044  [{'feature': 'Hour', 'attribution': -0.0287152...   
4             6.327058  [{'feature': 'Hour', 'attribution': 0.11369922...   
5             6.327058  [{'feature': 'Hour', 'attribution': 0.11369922...   
6             5.942420  [{'feature': 'Year', 'attribution': 0.10889728...   
7             5.942420  [{'feature': 'Year', 'attribution': 0.10889728...   
8             5.955779  [{'feature': 'Year', 'attribution': 0.10763874...   
9             5.955779  [{'feature': 'Year', 'attribution': 0.10763874...   

   baseline_prediction_value  prediction_value  approximation_error  Latitude  \
0                   5.845811          6.068206                  0.0    19.250   
1                   5.845811          6.068206                  0.0    19.250   
2                   5.845811          5.822044                  0.0    42.126   
3                   5.845811          5.822044                  0.0    42.126   
4                   5.845811          6.327059                  0.0     6.405   
5                   5.845811          6.327059                  0.0     6.405   
6                   5.845811          5.942421                  0.0    44.639   
7                   5.845811          5.942421                  0.0    44.639   
8                   5.845811          5.955780                  0.0    -6.134   
9                   5.845811          5.955780                  0.0    -6.134   

   Longitude  Depth        Type Magnitude_Type  ...  Month  Day  Hour  \
0   -155.490   10.9  Earthquake            MWC  ...      4   17     0   
1   -155.490   10.9  Earthquake            MWC  ...      4   17     0   
2    135.008  395.0  Earthquake             MW  ...     12   24     0   
3    135.008  395.0  Earthquake             MW  ...     12   24     0   
4    126.640   60.0  Earthquake             MW  ...     12    2     0   
5    126.640   60.0  Earthquake             MW  ...     12    2     0   
6    149.276   39.0  Earthquake             MW  ...     12   17     0   
7    149.276   39.0  Earthquake             MW  ...     12   17     0   
8    148.846   35.0  Earthquake             MW  ...     12   19     0   
9    148.846   35.0  Earthquake             MW  ...     12   19     0   

      sin_month  cos_month  sin_hour  cos_hour  latitude_x_longitude  \
0  8.660254e-01       -0.5       0.0       1.0          -2993.182500   
1  8.660254e-01       -0.5       0.0       1.0          -2993.182500   
2 -2.449294e-16        1.0       0.0       1.0           5687.347008   
3 -2.449294e-16        1.0       0.0       1.0           5687.347008   
4 -2.449294e-16        1.0       0.0       1.0            811.129200   
5 -2.449294e-16        1.0       0.0       1.0            811.129200   
6 -2.449294e-16        1.0       0.0       1.0           6663.531364   
7 -2.449294e-16        1.0       0.0       1.0           6663.531364   
8 -2.449294e-16        1.0       0.0       1.0           -913.021364   
9 -2.449294e-16        1.0       0.0       1.0           -913.021364   

   depth_squared  Magnitude  
0         118.81        5.7  
1         118.81        5.7  
2      156025.00        5.7  
3      156025.00        5.7  
4        3600.00        8.0  
5        3600.00        8.0  
6        1521.00        6.4  
7        1521.00        6.4  
8        1225.00        5.8  
9        1225.00        5.8  

[10 rows x 21 columns]

# Interactive Plotly Chart

In [60]:
import plotly.express as px
import pandas as pd
import pandas_gbq
import os

PROJECT_ID = os.environ["GOOGLE_CLOUD_PROJECT"]
DATASET = "earthquake"
BQML_TABLE_NAME = "earthquake_bqml_enhanced_data"
bqml_table_id = f"{PROJECT_ID}.{DATASET}.{BQML_TABLE_NAME}"

# Load the enhanced data from BigQuery into a pandas DataFrame
df_enhanced = pandas_gbq.read_gbq(f"SELECT * FROM `{bqml_table_id}`", project_id=PROJECT_ID)

print("Shape of enhanced DataFrame:", df_enhanced.shape)
display(df_enhanced.head())

Downloading: 100%|██████████|
Shape of enhanced DataFrame: (46812, 23)


Latitude  Longitude  Depth        Type Magnitude Type  Year  Month  Day  \
0    19.250   -155.490   10.9  Earthquake            MWC  1999      4   17   
1    19.250   -155.490   10.9  Earthquake            MWC  1999      4   17   
2    42.126    135.008  395.0  Earthquake             MW  2009     12   24   
3    42.126    135.008  395.0  Earthquake             MW  2009     12   24   
4     6.405    126.640   60.0  Earthquake             MW  1972     12    2   

   Hour              ID  ...     Status Magnitude     sin_month cos_month  \
0     0    GCMTB041799A  ...  Automatic       5.7  8.660254e-01      -0.5   
1     0    GCMTB041799A  ...  Automatic       5.7  8.660254e-01      -0.5   
2     0  ISCGEM17576175  ...  Automatic       5.7 -2.449294e-16       1.0   
3     0  ISCGEM17576175  ...  Automatic       5.7 -2.449294e-16       1.0   
4     0    ISCGEM765501  ...  Automatic       8.0 -2.449294e-16       1.0   

   sin_hour  cos_hour  latitude_x_longitude  depth_squared  depth_x_magnitude  \
0       0.0       1.0          -2993.182500         118.81              62.13   
1       0.0       1.0          -2993.182500         118.81              62.13   
2       0.0       1.0           5687.347008      156025.00            2251.50   
3       0.0       1.0           5687.347008      156025.00            2251.50   
4       0.0       1.0            811.129200        3600.00             480.00   

   magnitude_squared  
0              32.49  
1              32.49  
2              32.49  
3              32.49  
4              64.00  

[5 rows x 23 columns]

In [61]:
# Create an interactive scatter plot of Magnitude vs Depth
fig = px.scatter(
    df_enhanced,
    x="Depth",
    y="Magnitude",
    color="Magnitude Type", # Color points by Magnitude Type for more insight
    hover_data=['Latitude', 'Longitude', 'Year', 'Month', 'Day', 'Hour'], # Show additional info on hover
    title="Earthquake Magnitude vs. Depth (Interactive)",
    labels={
        "Depth": "Earthquake Depth (km)",
        "Magnitude": "Earthquake Magnitude",
        "Magnitude Type": "Magnitude Type"
    },
    height=600
)

fig.update_layout(hovermode="closest")
fig.show()